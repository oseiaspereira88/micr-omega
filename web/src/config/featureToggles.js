const parseBooleanFlag = (value, defaultValue = false) => {
  if (value === undefined || value === null) {
    return defaultValue;
  }

  const normalized = String(value).trim().toLowerCase();
  if (normalized === "true" || normalized === "1" || normalized === "on") {
    return true;
  }
  if (normalized === "false" || normalized === "0" || normalized === "off") {
    return false;
  }

  return defaultValue;
};

export const featureToggles = Object.freeze({
  minimap: parseBooleanFlag(import.meta.env.VITE_FEATURE_MINIMAP, false),
});
