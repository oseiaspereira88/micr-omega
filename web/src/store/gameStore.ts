import { useSyncExternalStore } from "react";
import {
  GamePhase,
  HealthState,
  Microorganism,
  OrganicMatter,
  Obstacle,
  OrientationState,
  RankingEntry,
  RoomObject,
  SharedGameState,
  SharedGameStateDiff,
  SharedPlayerState,
  SharedWorldState,
  Vector2,
} from "../utils/messageTypes";
import { reportRealtimeLatency } from "../utils/observability";

export type ConnectionStatus =
  | "idle"
  | "connecting"
  | "connected"
  | "reconnecting"
  | "disconnected";

export type PlayerMap = Record<string, SharedPlayerState>;

export type EntityCollection<T extends { id: string }> = {
  byId: Record<string, T>;
  all: T[];
};

type WorldCollections = {
  microorganisms: EntityCollection<Microorganism>;
  organicMatter: EntityCollection<OrganicMatter>;
  obstacles: EntityCollection<Obstacle>;
  roomObjects: EntityCollection<RoomObject>;
};

const createEmptyEntityCollection = <T extends { id: string }>(): EntityCollection<T> => ({
  byId: {},
  all: [],
});

const createEntityCollectionFromArray = <T extends { id: string }>(
  items: readonly T[],
  cloneItem: (item: T) => T,
): EntityCollection<T> => {
  const all = items.map((item) => cloneItem(item));
  const byId: Record<string, T> = {};
  for (const item of all) {
    byId[item.id] = item;
  }
  return { byId, all };
};

const applyEntityCollectionDiff = <T extends { id: string }>(
  collection: EntityCollection<T>,
  upserts: readonly T[] | undefined,
  removals: readonly string[] | undefined,
  cloneItem: (item: T) => T,
): { next: EntityCollection<T>; changed: boolean } => {
  let nextAll = collection.all;
  let nextById = collection.byId;
  let changed = false;

  if (upserts && upserts.length > 0) {
    nextAll = collection.all.slice();
    nextById = { ...collection.byId };
    const indexById = new Map<string, number>();
    nextAll.forEach((item, index) => {
      indexById.set(item.id, index);
    });

    for (const entry of upserts) {
      const cloned = cloneItem(entry);
      const existingIndex = indexById.get(cloned.id);
      if (existingIndex !== undefined) {
        nextAll[existingIndex] = cloned;
      } else {
        indexById.set(cloned.id, nextAll.push(cloned) - 1);
      }
      nextById[cloned.id] = cloned;
    }

    changed = true;
  }

  if (removals && removals.length > 0) {
    const removalSet = new Set(removals);
    if (removalSet.size > 0) {
      if (!changed) {
        nextAll = collection.all.slice();
        nextById = { ...collection.byId };
      }

      const filtered: T[] = [];
      let removedAny = false;
      for (const item of nextAll) {
        if (removalSet.has(item.id)) {
          delete nextById[item.id];
          removedAny = true;
        } else {
          filtered.push(item);
        }
      }

      if (removedAny) {
        nextAll = filtered;
        changed = true;
      }
    }
  }

  if (!changed) {
    return { next: collection, changed: false };
  }

  return {
    next: {
      byId: nextById,
      all: nextAll,
    },
    changed: true,
  };
};

const createEmptyWorldCollections = (): WorldCollections => ({
  microorganisms: createEmptyEntityCollection<Microorganism>(),
  organicMatter: createEmptyEntityCollection<OrganicMatter>(),
  obstacles: createEmptyEntityCollection<Obstacle>(),
  roomObjects: createEmptyEntityCollection<RoomObject>(),
});

const buildWorldFromCollections = (collections: WorldCollections): SharedWorldState => ({
  microorganisms: collections.microorganisms.all,
  organicMatter: collections.organicMatter.all,
  obstacles: collections.obstacles.all,
  roomObjects: collections.roomObjects.all,
});

export interface RoomStateSnapshot {
  phase: GamePhase;
  roundId: string | null;
  roundStartedAt: number | null;
  roundEndsAt: number | null;
}

export interface GameStoreState {
  connectionStatus: ConnectionStatus;
  reconnectAttempts: number;
  reconnectUntil: number | null;
  playerId: string | null;
  playerName: string | null;
  joinError: string | null;
  lastPingAt: number | null;
  lastPongAt: number | null;
  room: RoomStateSnapshot;
  players: PlayerMap;
  remotePlayers: EntityCollection<SharedPlayerState>;
  ranking: RankingEntry[];
  microorganisms: EntityCollection<Microorganism>;
  organicMatter: EntityCollection<OrganicMatter>;
  obstacles: EntityCollection<Obstacle>;
  roomObjects: EntityCollection<RoomObject>;
  world: SharedWorldState;
}

const defaultRoomState: RoomStateSnapshot = {
  phase: "waiting",
  roundId: null,
  roundStartedAt: null,
  roundEndsAt: null,
};

const createEmptySynchronizedState = () => {
  const remotePlayers = createEmptyEntityCollection<SharedPlayerState>();
  const worldCollections = createEmptyWorldCollections();
  return {
    remotePlayers,
    microorganisms: worldCollections.microorganisms,
    organicMatter: worldCollections.organicMatter,
    obstacles: worldCollections.obstacles,
    roomObjects: worldCollections.roomObjects,
    world: buildWorldFromCollections(worldCollections),
  };
};

const emptySyncState = createEmptySynchronizedState();

const initialState: GameStoreState = {
  connectionStatus: "idle",
  reconnectAttempts: 0,
  reconnectUntil: null,
  playerId: null,
  playerName: null,
  joinError: null,
  lastPingAt: null,
  lastPongAt: null,
  room: defaultRoomState,
  players: emptySyncState.remotePlayers.byId,
  remotePlayers: emptySyncState.remotePlayers,
  ranking: [],
  microorganisms: emptySyncState.microorganisms,
  organicMatter: emptySyncState.organicMatter,
  obstacles: emptySyncState.obstacles,
  roomObjects: emptySyncState.roomObjects,
  world: emptySyncState.world,
};

type GameStoreListener = () => void;

type StateUpdater = (state: GameStoreState) => GameStoreState;

type StateSelector<T> = (state: GameStoreState) => T;

let currentState: GameStoreState = initialState;
const listeners = new Set<GameStoreListener>();

const notify = () => {
  listeners.forEach((listener) => listener());
};

const applyState = (updater: StateUpdater) => {
  const nextState = updater(currentState);
  if (nextState !== currentState) {
    currentState = nextState;
    notify();
  }
};

const cloneVector = (vector: Vector2): Vector2 => ({ x: vector.x, y: vector.y });

const cloneOrientation = (orientation: OrientationState): OrientationState =>
  orientation.tilt === undefined
    ? { angle: orientation.angle }
    : { angle: orientation.angle, tilt: orientation.tilt };

const cloneHealth = (health: HealthState): HealthState => ({
  current: health.current,
  max: health.max,
});

const cloneCombatStatus = (
  status: SharedPlayerState["combatStatus"],
): SharedPlayerState["combatStatus"] => ({
  state: status.state,
  targetPlayerId: status.targetPlayerId,
  targetObjectId: status.targetObjectId,
  lastAttackAt: status.lastAttackAt,
});

const cloneCombatAttributes = (
  attributes: SharedPlayerState["combatAttributes"],
): SharedPlayerState["combatAttributes"] => ({
  attack: attributes.attack,
  defense: attributes.defense,
  speed: attributes.speed,
  range: attributes.range,
});

const clonePlayer = (player: SharedPlayerState): SharedPlayerState => ({
  ...player,
  position: cloneVector(player.position),
  movementVector: cloneVector(player.movementVector),
  orientation: cloneOrientation(player.orientation),
  health: cloneHealth(player.health),
  combatStatus: cloneCombatStatus(player.combatStatus),
  combatAttributes: cloneCombatAttributes(player.combatAttributes),
});

const cloneMicroorganism = (entity: Microorganism): Microorganism => ({
  ...entity,
  position: cloneVector(entity.position),
  movementVector: cloneVector(entity.movementVector),
  orientation: cloneOrientation(entity.orientation),
  health: cloneHealth(entity.health),
  attributes: { ...entity.attributes },
});

const cloneOrganicMatter = (matter: OrganicMatter): OrganicMatter => ({
  ...matter,
  position: cloneVector(matter.position),
  nutrients: { ...matter.nutrients },
});

const cloneObstacle = (obstacle: Obstacle): Obstacle => ({
  ...obstacle,
  position: cloneVector(obstacle.position),
  size: cloneVector(obstacle.size),
  orientation: obstacle.orientation ? cloneOrientation(obstacle.orientation) : undefined,
});

const cloneRoomObject = (object: RoomObject): RoomObject => ({
  ...object,
  position: cloneVector(object.position),
  state: object.state ? { ...object.state } : undefined,
});

const isFullState = (
  state: SharedGameState | SharedGameStateDiff
): state is SharedGameState => Array.isArray((state as SharedGameState).players);

const deriveRoomFromState = (
  state: SharedGameState | SharedGameStateDiff,
  previousRoom: RoomStateSnapshot
): RoomStateSnapshot => {
  if (isFullState(state)) {
    return {
      phase: state.phase,
      roundId: state.roundId,
      roundStartedAt: state.roundStartedAt,
      roundEndsAt: state.roundEndsAt,
    };
  }

  const nextRoom: RoomStateSnapshot = { ...previousRoom };

  if (state.phase !== undefined) {
    nextRoom.phase = state.phase;
  }

  if (state.roundId !== undefined) {
    nextRoom.roundId = state.roundId;
  }

  if (state.roundStartedAt !== undefined) {
    nextRoom.roundStartedAt = state.roundStartedAt;
  }

  if (state.roundEndsAt !== undefined) {
    nextRoom.roundEndsAt = state.roundEndsAt;
  }

  return nextRoom;
};

const setConnectionStatus = (status: ConnectionStatus) => {
  applyState((prev) => {
    if (prev.connectionStatus === status) {
      return prev;
    }

    const next: GameStoreState = {
      ...prev,
      connectionStatus: status,
    };

    if (status === "connected") {
      next.reconnectAttempts = 0;
    }

    if (status === "disconnected") {
      next.reconnectUntil = null;
    }

    return next;
  });
};

const setPlayerName = (name: string | null) => {
  applyState((prev) => {
    if (prev.playerName === name) {
      return prev;
    }

    return {
      ...prev,
      playerName: name,
    };
  });
};

const setPlayerId = (playerId: string | null) => {
  applyState((prev) => {
    if (prev.playerId === playerId) {
      return prev;
    }

    return {
      ...prev,
      playerId,
    };
  });
};

const setJoinError = (message: string | null) => {
  applyState((prev) => {
    if (prev.joinError === message) {
      return prev;
    }

    return {
      ...prev,
      joinError: message,
    };
  });
};

const incrementReconnectAttempts = () => {
  let attempts = 0;
  applyState((prev) => {
    attempts = prev.reconnectAttempts + 1;
    return {
      ...prev,
      reconnectAttempts: attempts,
    };
  });
  return attempts;
};

const setReconnectUntil = (timestamp: number | null) => {
  applyState((prev) => {
    if (prev.reconnectUntil === timestamp) {
      return prev;
    }

    return {
      ...prev,
      reconnectUntil: timestamp,
    };
  });
};

const applyFullState = (state: SharedGameState) => {
  applyState((prev) => {
    const remotePlayers = createEntityCollectionFromArray(state.players, clonePlayer);
    const worldCollections: WorldCollections = {
      microorganisms: createEntityCollectionFromArray(
        state.world.microorganisms,
        cloneMicroorganism,
      ),
      organicMatter: createEntityCollectionFromArray(state.world.organicMatter, cloneOrganicMatter),
      obstacles: createEntityCollectionFromArray(state.world.obstacles, cloneObstacle),
      roomObjects: createEntityCollectionFromArray(state.world.roomObjects, cloneRoomObject),
    };

    return {
      ...prev,
      room: {
        phase: state.phase,
        roundId: state.roundId,
        roundStartedAt: state.roundStartedAt,
        roundEndsAt: state.roundEndsAt,
      },
      players: remotePlayers.byId,
      remotePlayers,
      microorganisms: worldCollections.microorganisms,
      organicMatter: worldCollections.organicMatter,
      obstacles: worldCollections.obstacles,
      roomObjects: worldCollections.roomObjects,
      world: buildWorldFromCollections(worldCollections),
    };
  });
};

const applyStateDiff = (diff: SharedGameStateDiff) => {
  applyState((prev) => {
    const nextRoom = deriveRoomFromState(diff, prev.room);

    const playersResult = applyEntityCollectionDiff(
      prev.remotePlayers,
      diff.upsertPlayers,
      diff.removedPlayerIds,
      clonePlayer,
    );

    const worldDiff = diff.world;
    const microorganismsResult = applyEntityCollectionDiff(
      prev.microorganisms,
      worldDiff?.upsertMicroorganisms,
      worldDiff?.removeMicroorganismIds,
      cloneMicroorganism,
    );
    const organicMatterResult = applyEntityCollectionDiff(
      prev.organicMatter,
      worldDiff?.upsertOrganicMatter,
      worldDiff?.removeOrganicMatterIds,
      cloneOrganicMatter,
    );
    const obstacleResult = applyEntityCollectionDiff(
      prev.obstacles,
      worldDiff?.upsertObstacles,
      worldDiff?.removeObstacleIds,
      cloneObstacle,
    );
    const roomObjectResult = applyEntityCollectionDiff(
      prev.roomObjects,
      worldDiff?.upsertRoomObjects,
      worldDiff?.removeRoomObjectIds,
      cloneRoomObject,
    );

    const worldChanged =
      microorganismsResult.changed ||
      organicMatterResult.changed ||
      obstacleResult.changed ||
      roomObjectResult.changed;

    const nextWorld = worldChanged
      ? buildWorldFromCollections({
          microorganisms: microorganismsResult.next,
          organicMatter: organicMatterResult.next,
          obstacles: obstacleResult.next,
          roomObjects: roomObjectResult.next,
        })
      : prev.world;

    return {
      ...prev,
      room: nextRoom,
      players: playersResult.next.byId,
      remotePlayers: playersResult.next,
      microorganisms: microorganismsResult.next,
      organicMatter: organicMatterResult.next,
      obstacles: obstacleResult.next,
      roomObjects: roomObjectResult.next,
      world: nextWorld,
    };
  });
};

const applyRanking = (ranking: RankingEntry[]) => {
  applyState((prev) => ({
    ...prev,
    ranking,
  }));
};

const markPing = (timestamp: number) => {
  applyState((prev) => ({
    ...prev,
    lastPingAt: timestamp,
  }));
};

const markPong = (timestamp: number) => {
  applyState((prev) => {
    const latency =
      typeof prev.lastPingAt === "number" ? Math.max(0, timestamp - prev.lastPingAt) : null;
    if (latency !== null) {
      reportRealtimeLatency(latency, { source: "ws", phase: prev.room.phase });
    }

    return {
      ...prev,
      lastPongAt: timestamp,
    };
  });
};

const resetGameState = () => {
  const preservedName = currentState.playerName;
  const preservedId = currentState.playerId;
  const preservedStatus = currentState.connectionStatus;
  const preservedReconnectUntil = currentState.reconnectUntil;
  applyState(() => {
    const emptyState = createEmptySynchronizedState();
    return {
      ...initialState,
      playerName: preservedName,
      playerId: preservedId,
      connectionStatus: preservedStatus,
      reconnectUntil: preservedReconnectUntil,
      players: emptyState.remotePlayers.byId,
      remotePlayers: emptyState.remotePlayers,
      microorganisms: emptyState.microorganisms,
      organicMatter: emptyState.organicMatter,
      obstacles: emptyState.obstacles,
      roomObjects: emptyState.roomObjects,
      world: emptyState.world,
    };
  });
};

export const gameStore = {
  getState: () => currentState,
  subscribe: (listener: GameStoreListener) => {
    listeners.add(listener);
    return () => {
      listeners.delete(listener);
    };
  },
  setState: applyState,
  setPartial: (partial: Partial<GameStoreState>) => {
    applyState((prev) => ({
      ...prev,
      ...partial,
    }));
  },
  actions: {
    setConnectionStatus,
    setPlayerName,
    setPlayerId,
    setJoinError,
    incrementReconnectAttempts,
    setReconnectUntil,
    applyFullState,
    applyStateDiff,
    applyRanking,
    markPing,
    markPong,
    resetGameState,
  },
};

export function useGameStore<T>(selector: StateSelector<T>): T {
  return useSyncExternalStore(
    gameStore.subscribe,
    () => selector(currentState),
    () => selector(initialState)
  );
}
