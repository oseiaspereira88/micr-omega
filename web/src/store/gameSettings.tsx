import {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
  type ReactNode,
} from "react";

import { featureToggles } from "../config/featureToggles.js";
import {
  DEFAULT_JOYSTICK_SENSITIVITY,
  DEFAULT_TOUCH_CONTROL_SCALE,
  clampJoystickSensitivity,
  clampTouchControlScale,
} from "../config/touchControls";

export type VisualDensity = "low" | "medium" | "high";

export type TouchLayout = "right" | "left";

export type GameSettings = {
  audioEnabled: boolean;
  masterVolume: number;
  visualDensity: VisualDensity;
  showTouchControls: boolean;
  showMinimap: boolean;
  touchLayout: TouchLayout;
  autoSwapTouchLayoutWhenSidebarOpen: boolean;
  touchControlScale: number;
  joystickSensitivity: number;
};

const DEFAULT_SETTINGS: GameSettings = {
  audioEnabled: true,
  masterVolume: 1,
  visualDensity: "medium",
  showTouchControls: false,
  showMinimap: featureToggles.minimap,
  touchLayout: "right",
  autoSwapTouchLayoutWhenSidebarOpen: true,
  touchControlScale: DEFAULT_TOUCH_CONTROL_SCALE,
  joystickSensitivity: DEFAULT_JOYSTICK_SENSITIVITY,
};

const STORAGE_KEY = "micr-omega:game-settings";

type GameSettingsContextValue = {
  settings: GameSettings;
  updateSettings: (next: Partial<GameSettings>) => void;
  resetSettings: () => void;
};

const GameSettingsContext = createContext<GameSettingsContextValue | undefined>(
  undefined
);

const parseBoolean = (value: unknown, fallback: boolean): boolean => {
  if (typeof value === "boolean") {
    return value;
  }

  if (typeof value === "string") {
    const normalized = value.trim().toLowerCase();
    if (normalized === "true") {
      return true;
    }
    if (normalized === "false") {
      return false;
    }
  }

  return fallback;
};

const parseVisualDensity = (value: unknown, fallback: VisualDensity): VisualDensity => {
  if (typeof value !== "string") {
    return fallback;
  }

  const normalized = value.trim().toLowerCase();
  if (normalized === "low" || normalized === "medium" || normalized === "high") {
    return normalized;
  }

  return fallback;
};

const parseTouchLayout = (value: unknown, fallback: TouchLayout): TouchLayout => {
  if (typeof value !== "string") {
    return fallback;
  }

  const normalized = value.trim().toLowerCase();
  if (normalized === "right" || normalized === "left") {
    return normalized;
  }

  return fallback;
};

const clampVolume = (value: number): number => {
  if (!Number.isFinite(value)) {
    return 1;
  }

  if (value < 0) {
    return 0;
  }

  if (value > 1) {
    return 1;
  }

  return value;
};

const parseMasterVolume = (value: unknown, fallback: number): number => {
  if (typeof value === "number") {
    return clampVolume(value);
  }

  if (typeof value === "string") {
    const parsed = Number.parseFloat(value);
    if (Number.isFinite(parsed)) {
      return clampVolume(parsed);
    }
  }

  return clampVolume(fallback);
};

const parseStoredSettings = (): GameSettings => {
  if (typeof window === "undefined") {
    return { ...DEFAULT_SETTINGS };
  }

  try {
    const raw = window.localStorage.getItem(STORAGE_KEY);
    if (!raw) {
      return { ...DEFAULT_SETTINGS };
    }

    const parsed = JSON.parse(raw) as Partial<GameSettings> | null;
    if (!parsed || typeof parsed !== "object") {
      return { ...DEFAULT_SETTINGS };
    }

    const audioEnabled = parseBoolean(parsed.audioEnabled, DEFAULT_SETTINGS.audioEnabled);

    const masterVolume = parseMasterVolume(parsed.masterVolume, DEFAULT_SETTINGS.masterVolume);

    const visualDensity = parseVisualDensity(
      parsed.visualDensity,
      DEFAULT_SETTINGS.visualDensity,
    );

    const showTouchControls = parseBoolean(
      parsed.showTouchControls,
      DEFAULT_SETTINGS.showTouchControls,
    );

    const showMinimap = parseBoolean(
      parsed.showMinimap,
      DEFAULT_SETTINGS.showMinimap,
    );

    const touchLayout = parseTouchLayout(
      parsed.touchLayout,
      DEFAULT_SETTINGS.touchLayout,
    );

    const autoSwapTouchLayoutWhenSidebarOpen = parseBoolean(
      parsed.autoSwapTouchLayoutWhenSidebarOpen,
      DEFAULT_SETTINGS.autoSwapTouchLayoutWhenSidebarOpen,
    );

    const touchControlScale = clampTouchControlScale(
      parsed.touchControlScale,
      DEFAULT_SETTINGS.touchControlScale,
    );

    const joystickSensitivity = clampJoystickSensitivity(
      parsed.joystickSensitivity,
      DEFAULT_SETTINGS.joystickSensitivity,
    );

    return {
      audioEnabled,
      masterVolume,
      visualDensity,
      showTouchControls,
      showMinimap,
      touchLayout,
      autoSwapTouchLayoutWhenSidebarOpen,
      touchControlScale,
      joystickSensitivity,
    };
  } catch (error) {
    console.warn("Não foi possível ler as configurações do jogo salvas", error);
    return { ...DEFAULT_SETTINGS };
  }
};

export const GameSettingsProvider = ({ children }: { children: ReactNode }) => {
  const [settings, setSettings] = useState<GameSettings>(parseStoredSettings);

  useEffect(() => {
    if (typeof window === "undefined") {
      return;
    }

    try {
      window.localStorage.setItem(STORAGE_KEY, JSON.stringify(settings));
    } catch (error) {
      console.warn("Não foi possível persistir as configurações do jogo", error);
    }
  }, [settings]);

  const updateSettings = useCallback((next: Partial<GameSettings>) => {
    setSettings((prev) => {
      const normalizedNext: Partial<GameSettings> = { ...next };

      if ("masterVolume" in normalizedNext && typeof normalizedNext.masterVolume === "number") {
        normalizedNext.masterVolume = clampVolume(normalizedNext.masterVolume);
      }

      if ("touchControlScale" in normalizedNext) {
        normalizedNext.touchControlScale = clampTouchControlScale(
          normalizedNext.touchControlScale,
          prev.touchControlScale,
        );
      }

      if ("joystickSensitivity" in normalizedNext) {
        normalizedNext.joystickSensitivity = clampJoystickSensitivity(
          normalizedNext.joystickSensitivity,
          prev.joystickSensitivity,
        );
      }

      return {
        ...prev,
        ...normalizedNext,
      };
    });
  }, []);

  const resetSettings = useCallback(() => {
    setSettings({ ...DEFAULT_SETTINGS });
  }, []);

  const value = useMemo(
    () => ({
      settings,
      updateSettings,
      resetSettings,
    }),
    [settings, updateSettings, resetSettings]
  );

  return (
    <GameSettingsContext.Provider value={value}>
      {children}
    </GameSettingsContext.Provider>
  );
};

export const useGameSettings = (): GameSettingsContextValue => {
  const context = useContext(GameSettingsContext);
  if (!context) {
    throw new Error("useGameSettings deve ser usado dentro de GameSettingsProvider");
  }
  return context;
};

export const getDefaultSettings = (): GameSettings => ({ ...DEFAULT_SETTINGS });
