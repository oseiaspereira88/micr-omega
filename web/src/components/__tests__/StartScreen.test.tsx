import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import { act, fireEvent, render, screen, waitFor } from "@testing-library/react";
import StartScreen from "../StartScreen";
import { GameSettingsProvider, useGameSettings } from "../../store/gameSettings";
import { gameStore, type GameStoreState } from "../../store/gameStore";
import { MIN_NAME_LENGTH } from "../../utils/messageTypes";
import { INVALID_PLAYER_NAME_MESSAGE } from "../../utils/playerNameStorage";
import React, { useEffect } from "react";
import { featureToggles } from "../../config/featureToggles.js";

const baseState = gameStore.getState();

const snapshot = (): GameStoreState => {
  const emptyPlayers = { byId: {}, all: [] };
  const emptyMicroorganisms = { byId: {}, all: [] };
  const emptyOrganic = { byId: {}, all: [] };
  const emptyObstacles = { byId: {}, all: [] };
  const emptyRoomObjects = { byId: {}, all: [] };

  return {
    ...baseState,
    room: { ...baseState.room },
    players: emptyPlayers.byId,
    remotePlayers: emptyPlayers,
    ranking: [],
    microorganisms: emptyMicroorganisms,
    organicMatter: emptyOrganic,
    obstacles: emptyObstacles,
    roomObjects: emptyRoomObjects,
    connectionStatus: "idle",
    playerId: null,
    playerName: null,
    joinError: null,
    world: {
      microorganisms: emptyMicroorganisms.all,
      organicMatter: emptyOrganic.all,
      obstacles: emptyObstacles.all,
      roomObjects: emptyRoomObjects.all,
    },
  };
};

const resetStore = () => {
  act(() => {
    gameStore.setState(() => snapshot());
  });
};

const renderWithProviders = (ui: React.ReactNode) =>
  render(<GameSettingsProvider>{ui}</GameSettingsProvider>);

describe("StartScreen", () => {
  beforeEach(() => {
    resetStore();
    if (typeof window !== "undefined") {
      window.localStorage.clear();
    }
  });

  afterEach(() => {
    resetStore();
    if (typeof window !== "undefined") {
      window.localStorage.clear();
    }
  });

  it("valida o nome antes de entrar na partida", () => {
    const onStart = vi.fn();
    renderWithProviders(<StartScreen onStart={onStart} onQuit={() => {}} />);

    const startButton = screen.getByRole("button", { name: /entrar na partida/i });
    fireEvent.click(startButton);

    expect(onStart).not.toHaveBeenCalled();
    expect(screen.getByRole("alert")).toHaveTextContent("Informe um nome");

    const input = screen.getByLabelText(/nome do jogador/i);
    fireEvent.change(input, { target: { value: "a".repeat(MIN_NAME_LENGTH - 1) } });
    fireEvent.click(startButton);

    expect(onStart).not.toHaveBeenCalled();
    const validationAlert = screen.getByRole("alert");
    expect(validationAlert).toHaveTextContent(INVALID_PLAYER_NAME_MESSAGE);
    expect(validationAlert.textContent).toBe(INVALID_PLAYER_NAME_MESSAGE);
  });

  it("mantém o foco no campo de nome quando a validação falha", () => {
    renderWithProviders(<StartScreen onStart={() => {}} onQuit={() => {}} />);

    const input = screen.getByLabelText(/nome do jogador/i);
    const startButton = screen.getByRole("button", { name: /entrar na partida/i });

    fireEvent.click(startButton);

    expect(input).toHaveFocus();
  });

  it("persiste o nome sanitizado e envia as configurações", () => {
    const onStart = vi.fn();
    renderWithProviders(<StartScreen onStart={onStart} onQuit={() => {}} />);

    const input = screen.getByLabelText(/nome do jogador/i);
    fireEvent.change(input, { target: { value: "  Alice  " } });

    const audioToggle = screen.getByLabelText(/som ligado/i);
    fireEvent.click(audioToggle); // desativa áudio
    expect(audioToggle).toHaveAccessibleName(/som desligado/i);

    const densitySelect = screen.getByLabelText(/densidade visual/i);
    fireEvent.change(densitySelect, { target: { value: "high" } });

    const touchLayoutSelect = screen.getByLabelText(/layout dos controles touch/i);
    fireEvent.change(touchLayoutSelect, { target: { value: "left" } });

    const touchToggle = screen.getByLabelText(/mostrar controles/i);
    fireEvent.click(touchToggle);

    const startButton = screen.getByRole("button", { name: /entrar na partida/i });
    fireEvent.click(startButton);

    expect(onStart).toHaveBeenCalledTimes(1);
    expect(onStart.mock.calls[0][0]).toMatchObject({
      name: "Alice",
      settings: {
        audioEnabled: false,
        visualDensity: "high",
        showTouchControls: true,
        showMinimap: featureToggles.minimap,
        touchLayout: "left",
      },
      autoJoinRequested: true,
    });

    expect(gameStore.getState().playerName).toBe("Alice");
    expect(window.localStorage.getItem("micr-omega:player-name")).toBe("Alice");
  });

  it("preenche o campo com o nome persistido sem atualizar o store", async () => {
    const persistedName = "Persistido";
    window.localStorage.setItem("micr-omega:player-name", persistedName);

    renderWithProviders(<StartScreen onStart={() => {}} onQuit={() => {}} />);

    const input = screen.getByLabelText(/nome do jogador/i);

    await waitFor(() => {
      expect(input).toHaveValue(persistedName);
    });

    expect(gameStore.getState().playerName).toBeNull();
  });

  it("mantém o nome persistido ao sair", () => {
    const onQuit = vi.fn();

    act(() => {
      gameStore.actions.setPlayerName("Carol");
      gameStore.actions.setPlayerId("player-1");
      gameStore.actions.setConnectionStatus("connected");
    });

    window.localStorage.setItem("micr-omega:player-name", "Carol");

    renderWithProviders(<StartScreen onStart={() => {}} onQuit={onQuit} />);

    const quitButton = screen.getByRole("button", { name: /desconectar/i });
    fireEvent.click(quitButton);

    expect(onQuit).toHaveBeenCalledTimes(1);
    expect(window.localStorage.getItem("micr-omega:player-name")).toBe("Carol");
  });

  it("não exibe o controle de desconexão quando não há sessão ativa", () => {
    renderWithProviders(<StartScreen onStart={() => {}} onQuit={() => {}} />);

    expect(
      screen.queryByRole("button", { name: /desconectar/i })
    ).not.toBeInTheDocument();
  });

  it("atualiza o store de configurações quando os toggles mudam", async () => {
    const settingsSpy = vi.fn();

    const SettingsObserver = ({ onChange }: { onChange: (value: ReturnType<typeof useGameSettings>["settings"]) => void }) => {
      const { settings } = useGameSettings();
      useEffect(() => {
        onChange(settings);
      }, [onChange, settings]);
      return null;
    };

    render(
      <GameSettingsProvider>
        <SettingsObserver onChange={settingsSpy} />
        <StartScreen onStart={() => {}} onQuit={() => {}} />
      </GameSettingsProvider>
    );

    const audioToggle = screen.getByLabelText(/som ligado/i);
    fireEvent.click(audioToggle);
    expect(audioToggle).toHaveAccessibleName(/som desligado/i);

    const densitySelect = screen.getByLabelText(/densidade visual/i);
    fireEvent.change(densitySelect, { target: { value: "low" } });

    const touchLayoutSelect = screen.getByLabelText(/layout dos controles touch/i);
    fireEvent.change(touchLayoutSelect, { target: { value: "left" } });

    const touchToggle = screen.getByLabelText(/mostrar controles/i);
    fireEvent.click(touchToggle);

    await waitFor(() => {
      expect(settingsSpy).toHaveBeenCalled();
      const lastCall = settingsSpy.mock.calls.at(-1);
      expect(lastCall?.[0]).toMatchObject({
        audioEnabled: false,
        visualDensity: "low",
        showTouchControls: true,
        showMinimap: featureToggles.minimap,
        touchLayout: "left",
      });
    });
  });
});
