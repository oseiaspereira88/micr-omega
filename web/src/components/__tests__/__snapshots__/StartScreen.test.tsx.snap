// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StartScreen > atualiza a prévia dos controles touch conforme a seleção 1`] = `
<div
  aria-hidden="true"
  class="_root_21c337 _touchLayoutPreview_a3c48b"
  data-selected-layout="left"
  data-testid="start-touch-layout-preview"
>
  <div
    class="_card_21c337 _cardLeft_21c337 _cardActive_21c337"
    data-active="true"
    data-layout="left"
  >
    <div
      class="_cardContent_21c337"
    >
      <div
        class="_stick_21c337"
      />
      <div
        class="_buttons_21c337 _buttonsLeft_21c337"
      >
        <span
          class="_button_21c337"
        />
        <span
          class="_button_21c337 _buttonSecondary_21c337"
        />
        <span
          class="_button_21c337 _buttonTertiary_21c337"
        />
      </div>
    </div>
  </div>
  <div
    class="_card_21c337 _cardRight_21c337 _cardInactive_21c337"
    data-active="false"
    data-layout="right"
  >
    <div
      class="_cardContent_21c337"
    >
      <div
        class="_stick_21c337"
      />
      <div
        class="_buttons_21c337 _buttonsRight_21c337"
      >
        <span
          class="_button_21c337"
        />
        <span
          class="_button_21c337 _buttonSecondary_21c337"
        />
        <span
          class="_button_21c337 _buttonTertiary_21c337"
        />
      </div>
    </div>
  </div>
</div>
`;
